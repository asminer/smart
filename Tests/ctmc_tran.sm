


// Simple test of transient analysis for CTMCs

real thry(real lambda, real t) := exp(-lambda * t);

real thry(real lambda, real t1, real t2) := (exp(-lambda*t1)-exp(-lambda*t2))/lambda;

real tof(int i) :=  case(i, null, 
                      1 : 0.25,
                      2 : 0.5,
                      3 : 1.0,
                      4 : 2.0,
                      5 : 4.0,
                      6 : 8.0,
                      7 : infinity
                    );

ctmc foo(real lambda) := {
  state a, b;
  init(b:1);
  arcs(b:a:lambda);

  for (int i in {1..7}) {
    real inb[i] := prob_at(in_state(b), tof(i));
  }

  for (int i in {1..3}, int j in {4..7}) {
    real accb[i][j] := prob_acc(in_state(b), tof(i), tof(j));
  }

};

// #Report + mc_solve

for (real L in {0.5, 0.75, 1.0, 1.5, 2.0}) {

    print("Measures for lambda = ", L, "\n");

    print("Time":-10, "measure":-15, "theory":-15, "\n");
    for (int i in {1..7}) {
      print(tof(i):-10, foo(L).inb[i]:-15:6, thry(L, tof(i)):-15:6, "\n");
    }
    print("\n");

    print("t1":-6, "t2":-10, "measure":-15, "theory":-15, "\n");
    for (int i in {1..3}, int j in {4..7}) {
      print(tof(i):-6, tof(j):-10, foo(L).accb[i][j]:-15, thry(L, tof(i), tof(j)):-15, "\n");
    }
    print("\n");
}
