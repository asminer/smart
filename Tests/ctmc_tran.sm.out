Measures for lambda = 0.5
Time      measure        theory         
0.25      0.882497       0.882497       
0.5       0.778801       0.778801       
1         0.606531       0.606531       
2         0.367879       0.367879       
4         0.135335       0.135335       
8         0.0183156      0.0183156      
infinity  0              0              

t1    t2        measure        theory         
0.25  2         1.02923        1.02923        
0.25  4         1.49432        1.49432        
0.25  8         1.72836        1.72836        
0.25  infinity  1.76499        1.76499        
0.5   2         0.821843       0.821843       
0.5   4         1.28693        1.28693        
0.5   8         1.52097        1.52097        
0.5   infinity  1.5576         1.5576         
1     2         0.477302       0.477302       
1     4         0.942391       0.942391       
1     8         1.17643        1.17643        
1     infinity  1.21306        1.21306        

Measures for lambda = 0.75
Time      measure        theory         
0.25      0.829029       0.829029       
0.5       0.687289       0.687289       
1         0.472367       0.472367       
2         0.22313        0.22313        
4         0.0497871      0.0497871      
8         0.00247875     0.00247875     
infinity  0              0              

t1    t2        measure        theory         
0.25  2         0.807865       0.807865       
0.25  4         1.03899        1.03899        
0.25  8         1.10207        1.10207        
0.25  infinity  1.10537        1.10537        
0.5   2         0.618879       0.618879       
0.5   4         0.850003       0.850003       
0.5   8         0.913081       0.913081       
0.5   infinity  0.916386       0.916386       
1     2         0.332315       0.332315       
1     4         0.563439       0.563439       
1     8         0.626517       0.626517       
1     infinity  0.629822       0.629822       

Measures for lambda = 1
Time      measure        theory         
0.25      0.778801       0.778801       
0.5       0.606531       0.606531       
1         0.367879       0.367879       
2         0.135335       0.135335       
4         0.0183156      0.0183156      
8         0.000335463    0.000335463    
infinity  0              0              

t1    t2        measure        theory         
0.25  2         0.643465       0.643465       
0.25  4         0.760485       0.760485       
0.25  8         0.778465       0.778465       
0.25  infinity  0.778801       0.778801       
0.5   2         0.471195       0.471195       
0.5   4         0.588215       0.588215       
0.5   8         0.606195       0.606195       
0.5   infinity  0.606531       0.606531       
1     2         0.232544       0.232544       
1     4         0.349564       0.349564       
1     8         0.367544       0.367544       
1     infinity  0.367879       0.367879       

Measures for lambda = 1.5
Time      measure        theory         
0.25      0.687289       0.687289       
0.5       0.472367       0.472367       
1         0.22313        0.22313        
2         0.0497871      0.0497871      
4         0.00247875     0.00247875     
8         6.14421e-06    6.14421e-06    
infinity  0              0              

t1    t2        measure        theory         
0.25  2         0.425001       0.425001       
0.25  4         0.45654        0.45654        
0.25  8         0.458189       0.458189       
0.25  infinity  0.458193       0.458193       
0.5   2         0.28172        0.28172        
0.5   4         0.313259       0.313259       
0.5   8         0.314907       0.314907       
0.5   infinity  0.314911       0.314911       
1     2         0.115562       0.115562       
1     4         0.147101       0.147101       
1     8         0.148749       0.148749       
1     infinity  0.148753       0.148753       

Measures for lambda = 2
Time      measure        theory         
0.25      0.606531       0.606531       
0.5       0.367879       0.367879       
1         0.135335       0.135335       
2         0.0183156      0.0183156      
4         0.000335463    0.000335463    
8         1.12535e-07    1.12535e-07    
infinity  0              0              

t1    t2        measure        theory         
0.25  2         0.294108       0.294108       
0.25  4         0.303098       0.303098       
0.25  8         0.303265       0.303265       
0.25  infinity  0.303265       0.303265       
0.5   2         0.174782       0.174782       
0.5   4         0.183772       0.183772       
0.5   8         0.18394        0.18394        
0.5   infinity  0.18394        0.18394        
1     2         0.0585098      0.0585098      
1     4         0.0674999      0.0674999      
1     8         0.0676676      0.0676676      
1     infinity  0.0676676      0.0676676      

